code/src/Reducers/todos.js

```diff
@@ -0,0 +1,174 @@
+import { createSlice } from "@reduxjs/toolkit";
+
+const initialState = {
+  items: [],
+  filters: [{ filterBy: "Reset" }],
+  sort: [{ sortBy: "" }],
+  category: "all",
+  updatedTodos: [],
+  updatefilters: [],
+  updateSort: [],
+  updateCategory: [],
+};
+
+export const todos = createSlice({
+  name: "todos",
+  initialState,
+  reducers: {
+    createNewTodo: (store, action) => {
+      const {
+        task,
+        time,
+        deadline,
+        category,
+        overdue,
+        isTimed,
+      } = action.payload;
+      store.items.push({
+        id: store.items.length + 1, //sets id
+        description: task, //task text
+        isComplete: false, //completed or not
+        completedAt: "", //time completed
+        hidden: false, //task shown or not (used with filterings/categories)
+        createdAt: time, //time created
+        deadline: deadline, //if theere is a due time on the task it's stored here
+        overdue: overdue,
+        isTimed: isTimed,
+        category: category, // if the task has a category it's stored here
+      });
+    },
+    toggleComplete: (store, action) => {
+      const { id, time } = action.payload;
+
+      store.updatedTodos = store.items.map((item) => {
+        if (item.id === id) {
+          item.isComplete = !item.isComplete;
+          item.completedAt = time;
+
+          /* if the filter is on reset or if the filter matches the tasks should show */
+          store.filters[0].filterBy === "Reset" ||
+          item[store.filters[0].filterBy] === store.filters[0].value
+            ? (item.hidden = false)
+            : (item.hidden = true);
+
+          return item;
+        } else {
+          return item;
+        }
+      });
+
+      store.items = store.updatedTodos;
+    },
+    categorize: (store, action) => {
+      store.category = action.payload;
+      store.updatedTodos = store.items.map((task) => {
+        if (store.category === "all" || task.category[0] === store.category) {
+          return { ...task, hidden: false };
+        }
+        return { ...task, hidden: true };
+      });
+
+      //update individial tasks
+      store.items = store.updatedTodos;
+    },
+    sortTodos: (store, action) => {
+      const { sortBy, order } = action.payload;
+      store.updatedTodos = store.items.map((task) => task);
+
+      //sort
+      if (order) {
+        if (sortBy === "isComplete") {
+          store.updatedTodos.sort(
+            (a, b) => Number(b[sortBy]) - Number(a[sortBy])
+          );
+        } else {
+          store.updatedTodos.sort((a, b) => (a[sortBy] > b[sortBy] ? 1 : -1));
+        }
+        //reverse sort
+      } else {
+        if (sortBy === "isComplete") {
+          store.updatedTodos.sort(
+            (a, b) => Number(a[sortBy]) - Number(b[sortBy])
+          );
+        } else {
+          store.updatedTodos.sort((a, b) => (b[sortBy] > a[sortBy] ? 1 : -1));
```

Comment:
the same line as 85... maybe it is a good candidate to create a separate function?